### DISCUSSION TITLE
Suggestions on how to manage db backfills when using prisma migrate

### DISCUSSION THREAD

**Discussion by arnmishra** [prisma/prisma#20284] [Q&A]
I saw https://github.com/prisma/migrate/issues/494 which seems similar to some of what we are looking for but it seems like those commands have since been deprecated. 

tldr; we are going to regularly have situations where we need to make changes to our postgres data model and backfill data into the new models we are creating. For example, let's say we are splitting a `User` model into 2 different models (`User` and `Admin`). In order to do that, I'd want to:
1. Create the `Admin` table
2. Migrate certain `Users` into the `Admin` table
3. Remove those users from the `Users` table
4. Remove the `isAdmin` field from the `Users` table

In order to do that, I would imagine I'll have 2 different db migrations (1 for step 1, 1 for step 4). However, what is the best practice for doing 2 and 3? Would it be to write a bunch of complex sql commands in the prisma generated migration files? 

What I'd like to do is write simple TS scripts using the prisma client to do 2 and 3 and have those run between steps 1 and 4. In order to do that, I guess the way would be to run:
1. `prisma db execute --file [generated_migration_file_1]` (if I ran `prisma migrate deploy` it would run both the migrations at once it seems)
2. run some sort of typescript script that I manage elsewhere to do steps 2 and 3
3. `prisma db execute --file [generated_migration_file_2]`

Is there a better way to do this? Are there examples of others doing something similar? Writing a tool to manage all of this during production deploys seems like it'll be quite messy.

Another example of what might need to be done is if we have an enum role on the `User` table and one of those "roles" is `ADMIN`, I might want to delete the `ADMIN` enum. in order to do that, I first have to delete all users on the `User` table with role `ADMIN`. in that case as well, I'd want to do the delete using prisma client followed by the generated postgresql migration of updating the enum

**Comment by ludralph** [comment]
Hi @arnmishra 👋 

Thank you for raising this question.

We don't natively support data migration yet but what you are trying to achieve is possible using the [Expand and Contract pattern](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern).

In this case, you would have to first `Expand` by running `Migrate` on your database using `prisma migrate dev`, write a script, run it against your database, and then `Contract` by removing the redundant columns. You can see an example of using the Expand and Contract pattern [here](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime)

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
We have this feature request that should make it easier what you are trying to achieve:

- https://github.com/prisma/prisma/issues/11810

I would recommend you to add a 👍  to the feature request so that our product team can prioritise it.

**Reply by nurul3101** [reply]
@arnmishra Checking to see if you had a chance to look at the response? Please let us know in case you still need any help. If not, please consider marking the discussion as answered/closed.

**Reply by arnmishra** [reply]
Hi, sorry about the delay. 

@nurul3101 that feature request is exactly what I'm looking for, I come from the land of TypeORM so having ORM based transformations would be extremely useful for complex migrations.

@ludralph thanks for the links, I'm essentially attempting to do the expand and contract pattern with the key issue here being step 3 from the document. In order to actually migrate the code like is done in the example, we'd have to write extremely complex SQL for some cases. Rather than that, I want to do the migration via the prisma ORM using typescript to make it less error prone (and possibly testable). 

It seems like given Prisma's lack of support for this, we'll have to either write ad hoc scripts for migrations that we manage ourselves and find a way to run against our db or we would have to write them in raw sql with the prisma migration flows. Neither options feels like a reasonable solution so I'm definitely not looking forward to figure out what to do here.

Are there any final suggestions? if not I'll mark the conversation as closed and we will try to figure out how to go forward with this.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by 0xdevalias** [comment]
> What I'd like to do is write simple TS scripts using the prisma client to do 2 and 3 and have those run between steps 1 and 4. In order to do that, I guess the way would be to run:
> 
> 1. `prisma db execute --file [generated_migration_file_1]` (if I ran `prisma migrate deploy` it would run both the migrations at once it seems)
> 2. run some sort of typescript script that I manage elsewhere to do steps 2 and 3
> 3. `prisma db execute --file [generated_migration_file_2]`
> 
> Is there a better way to do this? Are there examples of others doing something similar? Writing a tool to manage all of this during production deploys seems like it'll be quite messy.

Specifically on the "if I ran `prisma migrate deploy` it would run both the migrations at once it seems" part; in a previous workplace on a different ORM, when doing essentially the expand/contract pattern, we would basically have 'dependent pull requests' with these different migration steps, so that the `main` branch would only ever have one migration merged into it waiting to be run. That way we never needed to specify to only run a single migration but not others after it.

We would then run our scripts separately, which as a little bit of boilerplate wrapper around running a remote batch script type job that just used our ORM to manipulate the data as needed (in that case we were using ActiveRecord in Ruby on Rails)

Then finally once we were happy with the data migration batch job completing successfully, we would merge/apply the latter migration that performed the 'contract' part of the pattern; and then once we were happy with all of that, we would often create a new PR to remove the migration batch script as well.
